app/code/community/Bolt/Boltpay/Model/Order.php

```diff
@@ -285,49 +256,321 @@ public function createOrder($reference, $sessionQuoteId = null, $isAjaxRequest =
+                            $now = Mage::getModel('core/date')->gmtTimestamp('Today');
+
+                            if ( $toTime && $toTime < $now ) {
+                                throw new Bolt_Boltpay_OrderCreationException(
+                                    OCE::E_BOLT_DISCOUNT_CANNOT_APPLY,
+                                    OCE::E_BOLT_DISCOUNT_CANNOT_APPLY_TMPL_EXPIRED,
+                                    array($boltCoupon->reference)
+                                );
+                            }
+
+                            throw new Bolt_Boltpay_OrderCreationException(
+                                OCE::E_BOLT_DISCOUNT_CANNOT_APPLY,
+                                OCE::E_BOLT_DISCOUNT_CANNOT_APPLY_TMPL_GENERIC,
+                                array("Coupon criteria was not met.", $boltCoupon->reference)
+                            );
+                        }
+
+                    } else {
+                        throw new Bolt_Boltpay_OrderCreationException(
+                            OCE::E_BOLT_DISCOUNT_DOES_NOT_EXIST,
+                            OCE::E_BOLT_DISCOUNT_DOES_NOT_EXIST_TMPL,
+                            array($boltCoupon->reference)
+                        );
+                    }
+                }
+
+            }
+        }
+    }
+
+    /**
+     * Verifies that the expected totals stored on Bolt have not changed in the Magento order prior to order creation
+     *
+     * @param Mage_Sales_Model_Quote $immutableQuote    Copy of the Magento session quote used by Bolt
+     * @param object                 $transaction       The Bolt transaction object sent from the Bolt server
+     *
+     * @throws Bolt_Boltpay_OrderCreationException upon failure of price consistency validation
+     */
+    protected function validateTotals(Mage_Sales_Model_Order $immutableQuote, $transaction)
+    {
+        $magentoTotals = $immutableQuote->getTotals();
+
+        foreach ($transaction->order->cart->items as $boltCartItem) {
+
+            $cartItem = $immutableQuote->getItemById($boltCartItem->reference);
+            $boltPrice = (int)$boltCartItem->total_amount->amount;
+            $magentoRowPrice = (int) ( $cartItem->getRowTotalWithDiscount() * 100 );
+            $magentoCalculatedPrice = (int) round($cartItem->getCalculationPrice() * 100 * $cartItem->getQty());
+
+            if ( !in_array($boltPrice, [$magentoRowPrice, $magentoCalculatedPrice]) ) {
+                throw new Bolt_Boltpay_OrderCreationException(
+                    OCE::E_BOLT_ITEM_PRICE_HAS_BEEN_UPDATED,
+                    OCE::E_BOLT_ITEM_PRICE_HAS_BEEN_UPDATED_TMPL,
+                    array($cartItem->getProductId(), $boltPrice, $magentoCalculatedPrice)
+                );
+            }
+        }
+
+        /////////////////////////////////////////////////////////////////////////
+        /// Historically, we have honored a price tolerance of 1 cent on
```

Comment:
pre-auth is a great opportunity for us to do this right; can we be strict here?

Reply:
maybe we can ignore (with warning log) some element's mismatch as long as total matches

Reply:
> pre-auth is a great opportunity for us to do this right; can we be strict here?
One cent is fairly strict considering we have known rounding problems, but we can set the default to zero for new merchants.  Existing merchants should generally remain at 1 cent to not introduce any new noise with the upgrade.

> maybe we can ignore (with warning log) some element's mismatch as long as total matches
This is the current behavior barring the logging.  I've now added warning logs.  We ignore the difference for individual subtotals as long as it does not exceed the overall tolerance.  A final check is implemented for overall price totals here:
https://github.com/BoltApp/bolt-magento1/blob/548420e59342aa979b2f4667968bf614b46a3302/app/code/community/Bolt/Boltpay/Model/Order.php#L338-L409